realsense/plugin.cpp

```diff
@@ -0,0 +1,179 @@
+#include <librealsense2/rs.hpp> // Include RealSense Cross Platform API
+#include <opencv2/opencv.hpp>   // Include OpenCV API
+#include <mutex>
+
+// ILLIXR includes
+#include "common/threadloop.hpp"
+#include "common/switchboard.hpp"
+#include "common/data_format.hpp"
+
+using namespace ILLIXR;
+
+static constexpr int IMAGE_WIDTH = 640;
+static constexpr int IMAGE_HEIGHT = 480;
+static constexpr int FPS = 30;
+static constexpr int GYRO_RATE = 400; // 200 or 400
+static constexpr int ACCEL_RATE = 250; // 63 or 250
+
+typedef struct {
+	cv::Mat* img0;
+	cv::Mat* img1;
+	cv::Mat* rgb;
+	cv::Mat* depth;
+	int iteration;
+} cam_type;
+
+typedef struct {
+	rs2_vector* accel_data;
+	int iteration;
+} accel_type;
+
+class realsense : public plugin {
+public:
+	realsense(std::string name_, phonebook *pb_)
+	: plugin{name_, pb_}
+	, sb{pb->lookup_impl<switchboard>()}
+	, _m_imu_cam{sb->publish<imu_cam_type>("imu_cam")}
+	, _m_rgb_depth{sb->publish<rgb_depth_type>("rgb_depth")}
+	{
+		auto callback = [&](const rs2::frame& frame)
+    {
+				std::lock_guard<std::mutex> lock(mutex);
```

Comment:
```suggestion
				std::lock_guard<std::mutex> lock(mutex);
				// This lock guarantees that concurrent invocations of `callback` are serialized.
				// Even if the API does not invoke `callback` in parallel, this is still important for the memory-model.
				// Without this lock, prior invocations of `callback` are not necessarily "happens-before" ordered, so accessing persistent variables constitutes a data-race, which is undefined behavior in the C++ memory model.
```

Reply:
@mhuzai Make sure I explained this properly.

Reply:
I don't completely follow. The idea is that the RealSense thread will call `callback` every time it has a frame but there is no guarantee that the previous invocation of `callback` would have finished by then?